pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,587 @@
+typealias T = AStruct // Don't need to qualify as `A::AStruct`.
+```
+
+All Slice definitions must be contained within modules - Slice definitions cannot exist at global scope.
+If a Slice file contains no Slice definitions, it is legal to omit the module declaration.
+Otherwise, if a Slice file contains any Slice definitions, a module declaration is required,
+and any Slice definitions in the remainder of the file will be contained within the declared module.
+
+Module declarations must come before any Slice definitions and must come after any [file attributes][attribute] or [mode declarations][mode-declaration].
+There can be at most one module declaration per Slice file.
+
+```ebnf {% showTitle=false %}
+SliceFile
+    : SliceFilePrelude Module? Definition*
+    ;
+```
+
+It is legal for multiple Slice files to declare the same module.
+Slice definitions defined in those files will be in the same module, even if they're in different Slice files.
+
+For additional information on modules, see the [module][module-guide] page.
+
+### Primitive types
+
+Slice supports 17 primitive types. These types are always available - they're defined in any scope of any Slice file -
+and can be referenced with their keywords:
+
+```ebnf {% showTitle=false %}
+Primitive
+    : "bool"      | "int8"  | "uint8"  | "int16"    | "uint16"    | "int32"   | "uint32"  | "varint32"
+    | "varuint32" | "int64" | "uint64" | "varint62" | "varuint62" | "float32" | "float64" | "string"
+    | "AnyClass"
+    ;
+```
+
+For additional information on primitive types, see the [primitive type][primitive-type-guide] page.
+
+### Collection types
+
+Slice supports 2 built-in collection types: sequences and dictionaries.
+
+These types are generic over the kinds of elements they contain.
+Sequences require a single type parameter: the _element_ type.
+Dictionaries require two type parameters: the _key_ and _value_ types.
+
+Any Slice type can be specified for element and value types, but since dictionary keys must be comparable, only the following are valid key types:
+- `bool`
+- `string`
+- integral types (`int8`, `uint8`, `int16`, `uint16`, `int32`, `uint32`, `varint32`, `varuint32`, `int64`, `uint64`, `varint62`, `varuint62`)
+- enum types
+- custom types
+- compact structs whose fields are also valid key types
+
+Sequences and dictionaries can be referenced with their respective keywords (`Sequence` and `Dictionary`),
+but must have concrete types specified for their type parameters.
+
+```ebnf {% showTitle=false %}
+Sequence
+    : "Sequence" "<" TypeRef ">"
+    ;
+
+Dictionary
+    : "Dictionary" "<" TypeRef "," TypeRef ">"
+    ;
+```
+
+```slice
+Sequence<bool>
+Dictionary<uint8, string>
+```
+
+For additional information on sequences and dictionaries, see the [sequence][sequence-guide] and [dictionary][dictionary-guide] pages.
+
+### Struct types
+
+Struct definitions consist of the `struct` keyword, followed by an [identifier], and then the struct's body.
```

Comment:
Unlike our other pages, I don't think users will read this one linearly (from the top to the bottom).
They'll probably just jump to the section that's relevant to them, skipping anything before it.

So instead of a link appearing the first time on the _page_, I added a link the first time it appears in a _section_.

If they want to read about 'identifiers', making users guess where they first appear, and scroll up to it hurts navigation.